hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/IMapIndexScanPhysicalRel.java

```diff
@@ -0,0 +1,191 @@
+        this.indexFilter = indexFilter;
+        this.converterTypes = converterTypes;
+        this.indexExp = indexExp;
+        this.remainderExp = remainderExp;
+    }
+
+    public Expression<Boolean> filter(QueryParameterMetadata parameterMetadata) {
+        PlanNodeSchema schema = OptUtils.schema(getTable());
+        return convertRemainderExpToFilter(schema, remainderExp, parameterMetadata);
+    }
+
+    public List<Expression<?>> projection(QueryParameterMetadata parameterMetadata) {
+        PlanNodeSchema schema = OptUtils.schema(getTable());
+
+        HazelcastTable table = getTable().unwrap(HazelcastTable.class);
+
+        List<Integer> projects = table.getProjects();
+        List<RexNode> projection = new ArrayList<>(projects.size());
+        for (Integer index : projects) {
+            TableField field = table.getTarget().getField(index);
+            RelDataType relDataType = OptUtils.convert(field, getCluster().getTypeFactory());
+            projection.add(new RexInputRef(index, relDataType));
+        }
+
+        return project(schema, projection, parameterMetadata);
+    }
+
+    @Override
+    public PlanNodeSchema schema(QueryParameterMetadata parameterMetadata) {
+        List<QueryDataType> fieldTypes = toList(projection(parameterMetadata), Expression::getType);
+        return new PlanNodeSchema(fieldTypes);
+    }
+
+    @Override
+    public Vertex accept(CreateDagVisitor visitor) {
+        return visitor.onIndexScan(this);
+    }
+
+    public MapTableIndex getIndex() {
+        return index;
+    }
+
+    public IndexFilter getIndexFilter() {
+        return indexFilter;
+    }
+
+    public List<QueryDataType> getConverterTypes() {
+        return converterTypes;
+    }
+
+    public List<Boolean> getAscs() {
+        RelCollation collation = getTraitSet().getTrait(RelCollationTraitDef.INSTANCE);
+        assert collation != null;
+        int size = collation.getFieldCollations().size();
+
+        List<Boolean> ascs = new ArrayList<>(size);
+        for (RelFieldCollation fieldCollation : collation.getFieldCollations()) {
+            Boolean asc = fieldCollation.getDirection() == Direction.ASCENDING ? TRUE : FALSE;
+            ascs.add(asc);
+        }
+        return ascs;
+    }
+
+    @Override
+    public RelNode copy(RelTraitSet traitSet, List<RelNode> inputs) {
+        return new IMapIndexScanPhysicalRel(
+                getCluster(),
+                traitSet,
+                getTable(),
+                index,
+                indexFilter,
+                converterTypes,
+                indexExp,
+                remainderExp
+        );
+    }
+
+    @Override
+    public RelWriter explainTerms(RelWriter pw) {
+        return super.explainTerms(pw)
+                .item("index", index.getName())
+                .item("indexExp", indexExp)
+                .item("remainderExp", remainderExp);
+    }
+
+    @Override
+    public double estimateRowCount(RelMetadataQuery mq) {
```

Comment:
I can see `MapIndexScanPhysicalRel` overrides `computeSelfCost()`, we don't need that?